drivers/powervar-hid.c

```diff
@@ -0,0 +1,146 @@
+/* powervar-hid.c - subdriver to monitor Powervar USB/HID devices with NUT
+ *
+ *  Copyright (C)
+ *  2003 - 2012	Arnaud Quette <ArnaudQuette@Eaton.com>
+ *  2005 - 2006	Peter Selinger <selinger@users.sourceforge.net>
+ *  2008 - 2009	Arjen de Korte <adkorte-guest@alioth.debian.org>
+ *  2013 Charles Lepple <clepple+nut@gmail.com>
+ *
+ *  Note: this subdriver was initially generated as a "stub" by the
+ *  gen-usbhid-subdriver script. It must be customized.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+#include "usbhid-ups.h"
+#include "powervar-hid.h"
+#include "main.h"	/* for getval() */
+#include "usb-common.h"
+
+#define POWERVAR_HID_VERSION	"Powervar HID 0.19"
+/* FIXME: experimental flag to be put in upsdrv_info */
+
+/* Powervar */
+#define POWERVAR_VENDORID	0x4234
+
+/* USB IDs device table */
+static usb_device_id_t powervar_usb_device_table[] = {
+	/* Powervar */
+	{ USB_DEVICE(POWERVAR_VENDORID, 0x0002), NULL },
+
+	/* Terminating entry */
+	{ -1, -1, NULL }
+};
+
+
+/* --------------------------------------------------------------- */
+/*      Vendor-specific usage table */
+/* --------------------------------------------------------------- */
+
+/* POWERVAR usage table */
+static usage_lkp_t powervar_usage_lkp[] = {
+	{ "POWERVAR1",	0xff000001 },
+	{  NULL, 0 }
+};
+
+static usage_tables_t powervar_utab[] = {
+	powervar_usage_lkp,
+	hid_usage_lkp,
+	NULL,
+};
+
+/* --------------------------------------------------------------- */
+/* HID2NUT lookup table                                            */
+/* --------------------------------------------------------------- */
+
+static hid_info_t powervar_hid2nut[] = {
+
+  { "ups.powersummary.iproduct", 0, 0, "UPS.PowerSummary.iProduct", NULL, "%.0f", 0, NULL },
```

Comment:
same comments as #807 on indexed string (mfr, model, serial)

Reply:
@aquette Are you saying that we don't need to have the iProduct, iSerialNumber & iManufacturer entries?

Reply:
Indeed

Reply:
Understood. I have removed them and diid a build and an install locally and all is well. I'll wait to upload until we have completed the other review items.

Reply:
Ok

Reply:
UPSC output with all changes included.

amccartney@amccartney-OptiPlex-3010:/lib/nut$ upsc UPM@localhost
Init SSL without certificate database
battery.capacity: 100
battery.capacitymode: 2
battery.charge: 100
battery.designcapacity: 100
battery.ioeminformation: AMETEK-POWERVAR Cor
battery.runtime: 360000
battery.type: PbAc
device.mfr: AMETEK-POWERVAR Corporation
device.model: UPM UPS (v01.51, Nov  3 2018)
device.serial: 5014403R-1740135
device.type: ups
driver.flag.pollonly: enabled
driver.name: usbhid-ups
driver.parameter.pollfreq: 30
driver.parameter.pollinterval: 2
driver.parameter.port: auto
driver.parameter.synchronous: no
driver.version: 2.7.4-732-gadb504d0
driver.version.data: Powervar HID 0.19
driver.version.internal: 0.43
ups.mfr: AMETEK-POWERVAR Corporation
ups.model: UPM UPS (v01.51, Nov  3 2018)
ups.powersummary.imanufacturer: 2
ups.powersummary.iproduct: 2
ups.powersummary.iserialnumber: 4
ups.productid: 0002
ups.serial: 5014403R-1740135
ups.status: OL
ups.vendorid: 4234

Reply:
thx @andrewmccartney

few more comments in light of this (@jimklimov  @clepple  and others may complete):

* battery.capacitymode: 2
** capacitymode doesn't exist (as per https://github.com/networkupstools/nut/blob/master/docs/nut-names.txt#L350)
** battery.capacity.mode would be more appropriate at first sight (new var, need RFC), but depends on what the values meaning
** for the value, there should be something more friendly to provide (2 == "some explicit string")

* battery.designcapacity
** battery.charge.design would be more appropriate (new var, need RFC), considering the value

* battery.ioeminformation
** non existing variable. battery.mfr or device.mfr.oem or ?

* device.model: UPM UPS (v01.51, Nov 3 2018)
** the part between parenthesis should be split between device.firmware and ups.mfr.date

* ups.powersummary.imanufacturer & ups.powersummary.iproduct & ups.powersummary.iserialnumber
** to be removed, as per my previous comment. these are just the indexes of the strings, already published as device.{mfr, model, serial}

* side comment: I see no load, realpower nor power data, that would give visibility on the power consumption on the UPS output. Nothing avail?

cheers,
Arno

Reply:
@aquette @jimklimov with regards to these changes, if I am understanding you correctly I would need to change the Powervar UPS HID behavior to accomplish this. This would be a UPS firmware change and at this time that is not a possibility. Please advise me if I am not understanding what you are asking here.

Reply:
@andrewmccartney no need to modify your FW for the above (possibly apart from the realpower, power and load Usage.
The other comments need just the NUT subdriver modification.
device.firmware could also be in its own "UPS.PowerSummary.iVersion", but can also be split from iModel using the callback function:
* modify powervar_format_model() so that it strips the parenthesis and just return the model
* leave a mapping on iProduct, but for device.firmware and set a callback function to extract the firmware info
{ "device.firmware", 0, 0, "UPS.PowerSummary.iProduct", NULL, "%.0f", 0, stringid_to_firmware},
* same for ups.mfr.date...